src/core/ext/transport/chttp2/transport/chttp2_transport.cc

```diff
@@ -1675,17 +1674,26 @@ class GracefulGoaway : public grpc_core::RefCounted<GracefulGoaway> {
   }
 
  private:
+  using TaskHandle = ::grpc_event_engine::experimental::EventEngine::TaskHandle;
+
   explicit GracefulGoaway(grpc_chttp2_transport* t) : t_(t) {
     t->sent_goaway_state = GRPC_CHTTP2_GRACEFUL_GOAWAY;
     GRPC_CHTTP2_REF_TRANSPORT(t_, "graceful goaway");
     grpc_chttp2_goaway_append((1u << 31) - 1, 0, grpc_empty_slice(), &t->qbuf);
     send_ping_locked(
         t, nullptr, GRPC_CLOSURE_INIT(&on_ping_ack_, OnPingAck, this, nullptr));
     grpc_chttp2_initiate_write(t, GRPC_CHTTP2_INITIATE_WRITE_GOAWAY_SENT);
-    Ref().release();  // Ref for the timer
-    grpc_timer_init(
-        &timer_, grpc_core::Timestamp::Now() + grpc_core::Duration::Seconds(20),
-        GRPC_CLOSURE_INIT(&on_timer_, OnTimer, this, nullptr));
+    timer_handle_ = t_->event_engine->RunAfter(
+        grpc_core::Duration::Seconds(20),
+        [self = Ref(DEBUG_LOCATION, "GoawayTimer")] {
+          grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+          grpc_core::ExecCtx exec_ctx;
+          self->Ref().release();
```

Comment:
something like `auto* ptr = self.release();` would be better and avoid the unnecessary Ref() and then Unref()?

Reply:
hmmm, I thought closures pushed onto the combiner might be got picked up and ran from another thread (not necessarily when the current exec_ctx destroys)? If that happens, we would want to take another Ref here. (and Unref in the combiner to keep the transport alive) Does my understanding correct?

Reply:
Great point! Just realize it (after the discussion) that if the timer closure is invoked, it's convenient to just release the RefCounterPtr in the closure and preserve the ref for the combiner (since the combiner will guarantee to be called).

If the timer closure is not invoked (e.g. cancelled), the RefCountedPtr will be destroyed when the closure is destroyed.